### PR TITLE
Pallavit/patchscript

### PR DESCRIPTION
**Pull request by pallavit**

### CODE REVIEW

File: eng/scripts/Generate-Patch.ps1 [hunk 1]
```diff
@@ -120,42 +106,45 @@ function UpdateChangeLog($ArtifactName, $ServiceDirectoryName, $Version) {
...
-  if('' -eq $PatchVersion) {
-    LogError "Could not fetch the patch version. Exitting ..."
-    exit
+function ResetSourcesToReleaseTag($ArtifactName, $ServiceDirectoryName, $ReleaseTag, $RepoRoot, $RemoteName) {
+  $SdkDirPath = Join-Path $RepoRoot "sdk"
+  $ServiceDirPath = Join-Path $SdkDirPath $ServiceDirectoryName
+  $ArtifactDirPath = Join-Path $ServiceDirPath $ArtifactName
+  $ArtifactPomFile = Join-Path $ArtifactDirPath "pom.xml"
+  
+  TestPathThrow -Path $ArtifactDirPath
```
Comment (alzimmermsft) [inline eng/scripts/Generate-Patch.ps1:115]: Should this be checked before joining with `pom.xml` to create `$ArtifactPomFile`

File: eng/scripts/Generate-Patch.ps1 [hunk 2]
```diff
@@ -165,24 +154,33 @@ try {
...
   if(Test-Path $SpotBugsFilePath) {
     $cmdOutput = git restore --source $ReleaseTag -W -S $SpotBugsFilePath
   }
 
-  ## Commit these changes.
+   ## Commit these changes.
   $cmdOutput = git commit -a -m "Reset changes to the patch version."
   if($LASTEXITCODE -ne 0) {
-    LogError "Could not commit the changes locally. to branch $BranchName. Exitting..."
+    LogError "Could not commit the changes locally.Exitting..."
```
Comment (alzimmermsft) [inline eng/scripts/Generate-Patch.ps1:169]: nit

```suggestion
    LogError "Could not commit the changes locally. Exiting..."
```

File: eng/scripts/Generate-Patch.ps1 [hunk 3]
```diff
@@ -120,42 +106,45 @@ function UpdateChangeLog($ArtifactName, $ServiceDirectoryName, $Version) {
...
+  }
+  
+  if($ResetSources -eq $false) {
+    return
+  }
+  
+  $TestResourcesFilePath = Join-Path $ServiceDirPath "test-resources.json"
+  $EngDir = Join-Path $RepoRoot "eng"  
+  $CodeQualityReports = Join-Path $EngDir "code-quality-reports" "src" "main" "resources"
+  $CheckStyleSuppressionFilePath = Join-Path $CodeQualityReports "checkstyle" "checkstyle-suppressions.xml"
```
Comment (alzimmermsft) [inline eng/scripts/Generate-Patch.ps1:138]: For Checkstyle you may need to reset the configuration in addition to the suppressions as the configurations could introduce new rules

File: eng/scripts/Generate-Patch.ps1 [hunk 4]
```diff
@@ -202,18 +200,44 @@ try {
...
-    if($LASTEXITCODE -ne 0) {
-      LogError "Could not checkout the branch $UpdateBranchName locally. Exitting..."
-      exit
-    }
+}
+
+if('' -eq $PatchVersion) {
+  $PatchVersion = GetPatchVersion -ReleaseVersion $ReleaseVersion
+  if('' -eq $PatchVersion) {
+    LogError "Could not fetch the patch version. Exitting ..."
```
Comment (alzimmermsft) [inline eng/scripts/Generate-Patch.ps1:208]:
```suggestion
    LogError "Could not fetch the patch version. Exiting..."
```

File: eng/scripts/Generate-Patch.ps1 [hunk 5]
```diff
@@ -202,18 +200,44 @@ try {
...
+    exit
+  }
+}
+Write-Information "PatchVersion is: $PatchVersion"
+
+$RemoteName = GetRemoteName -MainRemoteUrl $MainRemoteUrl
+if($null -eq $RemoteName) {
+    LogError "Could not fetch the remote name for the URL $MainRemoteUrl Exitting ..."
```
Comment (alzimmermsft) [inline eng/scripts/Generate-Patch.ps1:216]:
```suggestion
    LogError "Could not fetch the remote name for the URL $MainRemoteUrl. Exiting..."
```

File: eng/scripts/Generate-Patch.ps1 [hunk 6]
```diff
@@ -202,18 +200,44 @@ try {
...
+  $BranchName = "release/$ArtifactNameToLower/$PatchVersion"
+}
+
+try {
+  Write-Information "Reseting the $ArtifactName sources to the release $ReleaseTag."
+
+  ## Creating a new branch
+  $CmdOutput = git checkout -b $BranchName $RemoteName/main
+  if($LASTEXITCODE -ne 0) {
+    LogError "Could not checkout branch $BranchName, please check if it already exists and delete as necessary. Exitting..."
```
Comment (alzimmermsft) [inline eng/scripts/Generate-Patch.ps1:232]:
```suggestion
    LogError "Could not checkout branch $BranchName, please check if it already exists and delete as necessary. Exiting..."
```

File: eng/scripts/Generate-Patch.ps1 [hunk 7]
```diff
@@ -202,18 +200,44 @@ try {
...
   }
   
+  ## Hard reseting it to the contents of the release tag.
+  ## Fetching all the tags from the remote branch
+  ResetSourcesToReleaseTag -ArtifactName $ArtifactName -ServiceDirectoryName $ServiceDirectoryName -ReleaseTag $ReleaseTag -RepoRoot $RepoRoot -RemoteName $RemoteName
+  
+  CreatePatchRelease -ArtifactName $ArtifactName -ServiceDirectoryName $ServiceDirectoryName -PatchVersion $PatchVersion -RepoRoot $RepoRoot
   $cmdOutput = git add $RepoRoot
   if($LASTEXITCODE -ne 0) {
     LogError "Could not add the changes. Exitting..."
```
Comment (alzimmermsft) [inline eng/scripts/Generate-Patch.ps1:243]:
```suggestion
    LogError "Could not add the changes. Exiting..."
```

File: eng/scripts/Generate-Patch.ps1 [hunk 8]
```diff
@@ -202,18 +200,44 @@ try {
...
+}
+Write-Information "RemoteName is: $RemoteName"
+
+if('' -eq $BranchName) {
+  $ArtifactNameToLower = $ArtifactName.ToLower()
+  $BranchName = "release/$ArtifactNameToLower/$PatchVersion"
+}
+
+try {
+  Write-Information "Reseting the $ArtifactName sources to the release $ReleaseTag."
```
Comment (alzimmermsft) [inline eng/scripts/Generate-Patch.ps1:227]:
```suggestion
  Write-Information "Resetting the $ArtifactName sources to the release $ReleaseTag."
```

File: eng/scripts/Generate-Patch.ps1 (hunk 2 above)
Comment (pallavit) [reply eng/scripts/Generate-Patch.ps1:169]: Woops!

**Review by alzimmermsft — APPROVED** [review]

**Pull request merged**